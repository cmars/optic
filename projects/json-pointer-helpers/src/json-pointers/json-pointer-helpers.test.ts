import * as jsonPointer from 'json-pointer';
import jsonPointerHelpers from './json-pointer-helpers';

describe('json schema pointer assumptions', () => {
  it('can be trusted to escape symbols common in OpenAPI', () => {
    const result = jsonPointer.compile(['1', '/example/{todo}', '200']);
    expect(result).toMatchSnapshot();
  });
  it('compiling an empty array returns an empty path', () => {
    const result = jsonPointer.compile([]);
    expect(result).toMatchSnapshot();
  });

  it('decoding an empty string returns an empty path', () => {
    const result = jsonPointer.parse('');
    expect(result).toMatchSnapshot();
  });
});

describe('json schema pointer helpers', () => {
  it('appends are escaped', () => {
    expect(
      jsonPointerHelpers.append('/paths', '/example/{todoId}', 'get')
    ).toMatchSnapshot();
  });

  it('can cleanup stoplights mess', () => {
    const result = jsonPointerHelpers.unescapeUriSafePointer(
      '#/paths/~1pet~1%7BpetId%7D~1uploadImage/post/responses/200/content/application~1json/schema/properties'
    );
    expect(result).toMatchSnapshot();
  });
});

describe('relative', () => {
  it('can make pointer relative', () => {
    const a = jsonPointerHelpers.compile(['paths', '/example', 'schema']);
    const b = jsonPointerHelpers.compile([
      'paths',
      '/example',
      'schema',
      'start-trek',
      'voyager',
    ]);

    expect(jsonPointerHelpers.relative(b, a)).toMatchSnapshot();
  });

  it('throws if improper lineage make pointer relative', () => {
    const a = jsonPointerHelpers.compile(['paths', '/example', 'schema']);
    const b = jsonPointerHelpers.compile([
      'other-place',
      'start-trek',
      'voyager',
    ]);

    expect(() =>
      jsonPointerHelpers.relative(b, a)
    ).toThrowErrorMatchingSnapshot();
  });

  it('throws if improper lineage make pointer relative', () => {
    const a = jsonPointerHelpers.compile([
      'paths',
      '/example',
      'schema',
      'start-trek',
      'voyager',
      'season-1',
    ]);
    const b = jsonPointerHelpers.compile([
      'paths',
      '/example',
      'schema',
      'start-trek',
      'voyager',
    ]);

    expect(() =>
      jsonPointerHelpers.relative(b, a)
    ).toThrowErrorMatchingSnapshot();
  });
});
