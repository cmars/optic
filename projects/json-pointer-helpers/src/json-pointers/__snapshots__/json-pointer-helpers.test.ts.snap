// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`json schema pointer assumptions can be trusted to escape symbols common in OpenAPI 1`] = `"/1/~1example~1{todo}/200"`;

exports[`json schema pointer assumptions compiling an empty array returns an empty path 1`] = `""`;

exports[`json schema pointer assumptions decoding an empty string returns an empty path 1`] = `Array []`;

exports[`json schema pointer helpers appends are escaped 1`] = `"/paths/~1example~1{todoId}/get"`;

exports[`json schema pointer helpers can cleanup stoplights mess 1`] = `"#/paths/~1pet~1{petId}~1uploadImage/post/responses/200/content/application~1json/schema/properties"`;

exports[`relative can make pointer relative 1`] = `"/start-trek/voyager"`;

exports[`relative throws if improper lineage make pointer relative 1`] = `"/other-place/start-trek/voyager can not be relative to /paths/~1example/schema -- need same lineage"`;

exports[`relative throws if improper lineage make pointer relative 2`] = `"/paths/~1example/schema/start-trek/voyager can not be relative to /paths/~1example/schema/start-trek/voyager/season-1"`;
